node-tests/ast_plugins_test.js

```diff
@@ -0,0 +1,185 @@
+    input = yield createTempDir();
+    input.write(fixturify.readSync(`${__dirname}/fixtures`));
+    htmlbarsOptions = {
+      isHTMLBars: true,
+      templateCompiler: require('ember-source/dist/ember-template-compiler.js')
+    };
+  }));
+
+  afterEach(co.wrap(function*() {
+    if (tree) {
+      tree.unregisterPlugins();
+      yield tree.processor.processor._cache.clear();
+    }
+
+    if (builder) {
+      builder.cleanup();
+    }
+
+    yield input.dispose();
+
+    if (output) {
+      yield output.dispose();
+    }
+  }));
+
+  let rewriterCallCount;
+  function DivRewriterImpl(env) {
+    let rewriter = {
+      name: "test-div-rewriter",
+      tagNameFile: undefined,
+      tagName: "span",
+      resetDependencies() {
+        rewriter.tagNameFile = undefined;
+        rewriter.tagName = "span";
+      },
+      dependencies() {
+        return rewriter.tagNameFile ? [rewriter.tagNameFile] : [];
+      },
+      visitor: {
+        Program(node) {
+          let sourceFile = node.loc.source;
+          let pathInfo = sourceFile && path.parse(sourceFile);
+          if (pathInfo) {
+            if (pathInfo.base === "template.hbs") {
+              rewriterCallCount++;
+            }
+            let tagNameFile = input.path(`${pathInfo.name}.tagname`);
+            if (fs.existsSync(tagNameFile)) {
+              let tagName = fs.readFileSync(tagNameFile, "utf-8").trim();
+              rewriter.tagName = tagName;
+              rewriter.tagNameFile = tagNameFile;
+            }
+          }
+        },
+        ElementNode(node) {
+          if (node.tag === "div") {
+            node.tag = rewriter.tagName;
+          }
+        }
+      }
+    };
+    return rewriter;
+  }
+  const DivRewriter = addDependencyTracker(DivRewriterImpl, true);
+
+  they('are accepted and used.', co.wrap(function* () {
+    htmlbarsOptions.plugins = {
+      ast: [DivRewriter],
+    };
+
+    tree = new TemplateCompiler(input.path(), htmlbarsOptions);
+
+    output = createBuilder(tree);
+    yield output.build();
+
+    let templateOutput = output.readText('template.js');
+    assert.ok(!templateOutput.match(/div/));
+    assert.ok(templateOutput.match(/my-custom-element/));
+    assert.strictEqual(rewriterCallCount, 1);
+  }));
+
+  they('will bust the hot cache if the dependency changes.', co.wrap(function* () {
+    Object.assign(htmlbarsOptions, {
+      plugins: {
+        ast: [DivRewriter]
+      },
+      dependencyInvalidation: true,
+    });
+
+    tree = new TemplateCompiler(input.path(), htmlbarsOptions);
+
+    output = createBuilder(tree);
+    yield output.build();
+    let templateOutput = output.readText('template.js');
+    assert.ok(!templateOutput.match(/div/));
+    assert.ok(templateOutput.match(/my-custom-element/));
+    assert.strictEqual(rewriterCallCount, 1);
+
+    // The state didn't change. the output should be cached
+    // and the rewriter shouldn't be invoked.
+    yield output.build();
```

Comment:
Can you add an assertion asserting that there were no changes to the output (`assert.deepEqual(output.changes(), [])`)?